In [1]:
# Mounting Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Import libraries
import pandas as pd
import seaborn as sns
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.preprocessing import LabelEncoder
from sklearn.utils.class_weight import compute_sample_weight

# Read and Clean data


In [3]:
pitches = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/pitches.csv")
pd.set_option('display.max_columns', None)
pitches

/tmp/ipykernel_1664/1893180833.py:1: DtypeWarning: Columns (29,30) have mixed types. Specify dtype option on import or set low_memory=False.
  pitches = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/pitches.csv")


,uid,game_pk,year,date,team_id_b,team_id_p,inning,top,at_bat_num,pcount_at_bat,pcount_pitcher,balls,strikes,fouls,outs,is_final_pitch,final_balls,final_strikes,final_outs,start_tfs,start_tfs_zulu,batter_id,stand,b_height,pitcher_id,p_throws,at_bat_des,event,event2,event3,event4,away_team_runs,home_team_runs,score,pitch_des,pitch_id,type,pitch_tfs,pitch_tfs_zulu,x,y,sv_id,start_speed,end_speed,sz_top,sz_bot,pfx_x,pfx_z,px,pz,x0,z0,y0,vx0,vz0,vy0,ax,az,ay,break_length,break_y,break_angle,pitch_type,type_confidence,zone,nasty,spin_dir,spin_rate,cc,on_1b,on_2b,on_3b,runner1_id,runner1_start,runner1_end,runner1_event,runner1_score,runner1_rbi,runner1_earned,runner2_id,runner2_start,runner2_end,runner2_event,runner2_score,runner2_rbi,runner2_earned,runner3_id,runner3_start,runner3_end,runner3_event,runner3_score,runner3_rbi,runner3_earned,runner4_id,runner4_start,runner4_end,runner4_event,runner4_score,runner4_rbi,runner4_earned,runner5_id,runner5_start,runner5_end,runner5_event,runner5_score,runner5_rbi,runner5_earned,runner6_id,runner6_start,runner6_end,runner6_event,runner6_score,runner6_rbi,runner6_earned,runner7_id,runner7_start,runner7_end,runner7_event,runner7_score,runner7_rbi,runner7_earned,created_at,added_at,modified_at,modified_by
0,14143226,286874,2011,2011-03-31,108,118,1,1,1,1,1,0,0,0,0,0,2,1,1,201226,2011-03-31 20:12:26,430895,L,5-8,460024,R,"Maicer Izturis grounds out, second baseman Chr...",Groundout,NaN,NaN,NaN,0,0,NaN,Ball,3,B,201301.0,2011-03-31 20:13:01,105.58,180.46,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2016-03-03 21:33:20,2016-03-03 21:33:20,2016-03-03 21:33:20,1
1,14143227,286874,2011,2011-03-31,108,118,1,1,1,2,2,1,0,0,0,0,2,1,1,201226,2011-03-31 20:12:26,430895,L,5-8,460024,R,"Maicer Izturis grounds out, second baseman Chr...",Groundout,NaN,NaN,NaN,0,0,NaN,Ball,4,B,201319.0,2011-03-31 20:13:19,99.57,170.96,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2016-03-03 21:33:20,2016-03-03 21:33:20,2016-03-03 21:33:20,1
2,14143228,286874,2011,2011-03-31,108,118,1,1,1,3,3,2,0,0,0,0,2,1,1,201226,2011-03-31 20:12:26,430895,L,5-8,460024,R,"Maicer Izturis grounds out, second baseman Chr...",Groundout,NaN,NaN,NaN,0,0,NaN,Called Strike,5,S,201327.0,2011-03-31 20:13:27,95.28,152.83,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2016-03-03 21:33:20,2016-03-03 21:33:20,2016-03-03 21:33:20,1
3,14143229,286874,2011,2011-03-31,108,118,1,1,1,4,4,2,1,0,0,1,2,1,1,201226,2011-03-31 20:12:26,430895,L,5-8,460024,R,"Maicer Izturis grounds out, second baseman Chr...",Groundout,NaN,NaN,NaN,0,0,NaN,"In play, out(s)",6,X,180441.0,2011-03-31 18:04:41,93.56,168.37,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2016-03-03 21:33:20,2016-03-03 21:33:20,2016-03-03 21:33:20,1
4,14143230,286874,2011,2011-03-31,108,118,1,1,2,1,5,0,0,0,1,0,2,2,1,201354,2011-03-31 20:13:54,435062,R,5-10,460024,R,Howie Kendrick doubles (1) on a line drive to ...,Double,NaN,NaN,NaN,0,0,N

In [4]:
# Chronological daat to predict pitch
pitches = pitches.sort_values(by=['game_pk', 'inning', 'top', 'at_bat_num', 'pcount_at_bat'])

# Create the target variables and previous pitch
pitches['next_pitch_type'] = pitches.groupby(['game_pk', 'at_bat_num'])['pitch_type'].shift(-1)
pitches['previous_pitch_type'] = pitches.groupby(['game_pk', 'at_bat_num'])['pitch_type'].shift(1)

# Drop columns and data that may leak or is irrelevant
columns_to_drop = [
    'start_speed', 'end_speed', 'sz_top', 'sz_bot', 'pfx_x', 'pfx_z',
    'px', 'pz', 'x0', 'z0', 'y0', 'vx0', 'vz0', 'vy0', 'ax', 'az', 'ay',
    'break_length', 'break_y', 'break_angle', 'spin_dir', 'spin_rate',
    'type_confidence', 'zone', 'nasty',

    'type', 'event', 'event2', 'event3', 'event4', 'pitch_des',
    'at_bat_des', 'is_final_pitch', 'final_balls', 'final_strikes', 'final_outs',

    'uid', 'game_pk', 'year', 'team_id_b', 'team_id_p',
    'start_tfs', 'start_tfs_zulu', 'pitch_tfs', 'pitch_tfs_zulu',
    'sv_id', 'pitch_id', 'x', 'y', 'created_at', 'added_at',
    'modified_at', 'modified_by', 'at_bat_num'
]


# any errors in the list will be ignored
pitches_cleaned = pitches.drop(columns=columns_to_drop, errors='ignore')

runner_cols = [col for col in pitches_cleaned.columns if col.startswith('runner')]
pitches_cleaned = pitches_cleaned.drop(columns=runner_cols, errors='ignore')

# Drop original 'pitch_type' to prevent leakeage essentailly for the model not to teach
pitches_cleaned = pitches_cleaned.drop(columns=['pitch_type'], errors='ignore')

# Drop NaN/Na beofre training the model
pitches_cleaned = pitches_cleaned.dropna(subset=['next_pitch_type'])
print(pitches_cleaned.columns.tolist())

['date', 'inning', 'top', 'pcount_at_bat', 'pcount_pitcher', 'balls', 'strikes', 'fouls', 'outs', 'batter_id', 'stand', 'b_height', 'pitcher_id', 'p_throws', 'away_team_runs', 'home_team_runs', 'score', 'cc', 'on_1b', 'on_2b', 'on_3b', 'next_pitch_type', 'previous_pitch_type']


In [5]:
# Name the target groups(will help later to interpret results)
pitch_category_map = {
    # Fastballs (Hard)
    'FF': 'Fastball', 'FT': 'Fastball', 'SI': 'Fastball', 'FC': 'Fastball',
    # Breaking Balls (Movement)
    'SL': 'Breaking', 'CU': 'Breaking', 'KC': 'Breaking',
    # Off-speed (Deception)
    'CH': 'Offspeed', 'FS': 'Offspeed', 'KN': 'Offspeed', 'EP': 'Offspeed'
}

# Create new columns for the groups
pitches_cleaned['next_pitch_group'] = pitches_cleaned['next_pitch_type'].map(pitch_category_map)
pitches_cleaned['previous_pitch_group'] = pitches_cleaned['previous_pitch_type'].map(pitch_category_map)
pitches_cleaned = pitches_cleaned.dropna(subset=['next_pitch_group'])

In [6]:
pitches_cleaned

,date,inning,top,pcount_at_bat,pcount_pitcher,balls,strikes,fouls,outs,batter_id,stand,b_height,pitcher_id,p_throws,away_team_runs,home_team_runs,score,cc,on_1b,on_2b,on_3b,next_pitch_type,previous_pitch_type,next_pitch_group,previous_pitch_group
26,2011-03-31,1,0,1,1,0,0,0,0,449107,R,5-10,450308,R,0,0,NaN,NaN,NaN,NaN,NaN,FF,NaN,Fastball,NaN
27,2011-03-31,1,0,2,2,0,1,0,0,449107,R,5-10,450308,R,0,0,NaN,NaN,NaN,NaN,NaN,FF,FF,Fastball,Fastball
28,2011-03-31,1,0,3,3,1,1,0,0,449107,R,5-10,450308,R,0,0,NaN,NaN,NaN,NaN,NaN,FF,FF,Fastball,Fastball
29,2011-03-31,1,0,4,4,2,1,0,0,449107,R,5-10,450308,R,0,0,NaN,NaN,NaN,NaN,NaN,FF,FF,Fastball,Fastball
30,2011-03-31,1,0,5,5,2,2,1,0,449107,R,5-10,450308,R,0,0,NaN,NaN,NaN,NaN,NaN,SL,FF,Breaking,Fastball
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
718954,2011-10-28,9,1,1,5,0,0,0,1,435063,R,6-0,435400,R,2,6,NaN,NaN,NaN,NaN,NaN,FF,NaN,Fastball,NaN
718955,2011-10-28,9,1,2,6,0,1,0,1,435063,R,6-0,435400,R,2,6,NaN,NaN,NaN,NaN,NaN,FF,FF,Fastball,Fastball
718956,2011-10-28,9,1,3,7,1,1,0,1,435063,R,6-0,435400,R,2,6,NaN,NaN,NaN,NaN,NaN,FC,FF,Fastball,Fastball
718957,2011-10-28,9,1,4,8,1,2,1,1,435063,R,6-0,435400,R,2,6,NaN,NaN,NaN,NaN,NaN,FF,FF,Fastball,Fastball


# Prepare and Encode data

In [7]:
pitches_cleaned['previous_pitch_type'] = pitches_cleaned['previous_pitch_type'].fillna('First_Pitch')

# Convert base runner to binary values
for col in ['on_1b', 'on_2b', 'on_3b']:
    pitches_cleaned[col] = pitches_cleaned[col].notna().astype(int)

# Convert height to integers
def parse_height(height_str):
    if pd.isna(height_str):
        return 0
    parts = str(height_str).split('-')
    return int(parts[0]) * 12 + int(parts[1]) if len(parts) == 2 else 0

pitches_cleaned['b_height'] = pitches_cleaned['b_height'].apply(parse_height)

In [8]:
cols_to_drop_now = ['date', 'batter_id', 'pitcher_id']
pitches_cleaned = pitches_cleaned.drop(columns=cols_to_drop_now, errors='ignore')

In [9]:
categorical_cols = pitches_cleaned.select_dtypes(include=['object', 'category', 'string']).columns.tolist()

if 'next_pitch_type' in categorical_cols:
    categorical_cols.remove('next_pitch_type')
if 'next_pitch_group' in categorical_cols:
    categorical_cols.remove('next_pitch_group')

# Filter columns by 20 unique values
low_cardinality_cols = [col for col in categorical_cols if pitches_cleaned[col].nunique() < 20]
high_cardinality_cols = list(set(categorical_cols) - set(low_cardinality_cols))

print(f"Good: Safe to one-hot encode: {low_cardinality_cols}")
print(f"Bad: Dropping due to high cardinality (too many unique values): {high_cardinality_cols}")

# Get dummy variables
pitches_model = pd.get_dummies(pitches_cleaned, columns=low_cardinality_cols, drop_first=True)
pitches_model = pitches_model.drop(columns=high_cardinality_cols, errors='ignore')

# 5. Convert booleans to floats
pitches_model = pitches_model.astype(float, errors='ignore')


print("\nEncoding complete! New dataset shape:", pitches_model.shape)

Good: Safe to one-hot encode: ['stand', 'p_throws', 'score', 'previous_pitch_type', 'previous_pitch_group']
Bad: Dropping due to high cardinality (too many unique values): ['cc']

Encoding complete! New dataset shape: (524599, 38)


In [10]:
import gc
# Extract variables we want
y_granular = pitches_model['next_pitch_type']
y_grouped = pitches_model['next_pitch_group']

# Drop targets from the dataset
X = pitches_model.drop(columns=['next_pitch_type', 'next_pitch_group'])

del pitches_model
gc.collect()

# Split the data
# Using random_state=42 ensures X_train perfectly aligns with both y_train targets
X_train, X_test, y_train_gran, y_test_gran = train_test_split(X, y_granular, test_size=0.2, random_state=42)
_, _, y_train_grp, y_test_grp = train_test_split(X, y_grouped, test_size=0.2, random_state=42)

# Random Forest Model

In [11]:
# Model 1 with 17 pitch types
rf_granular = RandomForestClassifier(n_estimators=25, max_depth=15, class_weight='balanced', random_state=42, n_jobs=-1)
rf_granular.fit(X_train, y_train_gran)

preds_gran = rf_granular.predict(X_test)
print("\nGRANULAR MODEL RESULTS")
print("Accuracy:", accuracy_score(y_test_gran, preds_gran))
print(classification_report(y_test_gran, preds_gran))

# Model 2 The Grouped Model with 3 Categories
print("\nGrouped Model (3 Categories)...")
rf_grouped = RandomForestClassifier(n_estimators=25, max_depth=15, class_weight='balanced', random_state=42, n_jobs=-1)
rf_grouped.fit(X_train, y_train_grp)

preds_grp = rf_grouped.predict(X_test)
print("\nGROUPED MODEL RESULTS")
print("Accuracy:", accuracy_score(y_test_grp, preds_grp))
print(classification_report(y_test_grp, preds_grp))


GRANULAR MODEL RESULTS
Accuracy: 0.26871902401829967
              precision    recall  f1-score   support

          CH       0.22      0.25      0.23     11796
          CU       0.19      0.16      0.17      8112
          EP       0.00      0.00      0.00        27
          FC       0.34      0.23      0.27      6378
          FF       0.52      0.23      0.32     34218
          FS       0.06      0.32      0.10      1746
          FT       0.31      0.31      0.31     11437
          KC       0.06      0.23      0.10      1204
          KN       0.04      0.80      0.08       621
          SI       0.46      0.37      0.41     12248
          SL       0.30      0.31      0.31     17133

    accuracy                           0.27    104920
   macro avg       0.23      0.29      0.21    104920
weighted avg       0.37      0.27      0.29    104920


Grouped Model (3 Categories)...

GROUPED MODEL RESULTS
Accuracy: 0.416374380480366
              precision    recall  f1-score   sup

Granular and Grouped Models are the two models tha are the
classification reports that evaluates the model's predictive performance. The dataset contains a total of 104,920 pitches whhich is the support. The model above predicts the specific pitch types e.g., Four-Seam Fastball, Slider, Curveball.

Overall Accuracy: The model is highly inaccurate at this granular level for correctly predicting the exact pitch type only 26.9% of the time can it predict it correctly.

Class Imbalance: The data is heavily skewed as the Four-seam fastballs (FF) are the most common being 34,218 pitches, followed by Sliders (SL) with 17,133. In contrast rare pitches like the Eephus (EP) appear only 27 times.

Performance Extremes:
Best Precisions: The model is the most precise when guessing FF (0.52) and Sinkers/SI (0.46). This is evident and useful depending on the pitcher and teams played
Anomalies: The Knuckleball (KN) has an extremely low precision (0.04) but a highly disproportionate recall (0.80) and this would suggest the model over-predicts the knuckleball drastically by catching most of them and generating a massive number of false positives. The model completely failed to predict the Eephus (EP) showing 0.00 across the board.

Grouped Model Results (3 Categories)
To improve the models performance the pitch types were grouped  into three broader categories Breaking, Fastball, and Offspeed.

Overall Accuracy: By grouping the pitches into 3 categories the model's accuracy improves significantly to 41.6%
Category Breakdown:
Fastball: Represents the majority of pitches at 64,281 and in baseball this is expe ted. The model has strong precision here of 0.71 but struggles with recall at 0.36 meaning when it guesses a Fastball it is usually is right but it misses a large portion of actual fastballs.

Breaking: Contains 26,449 pitches in this category it is moderately balanced with a 0.35 precision and 0.50 recall.

Offspeed: Contains 14,190 pitches and the model has poor precision of 0.22 but captures slightly more than half of the actual offspeed pitches with a recall of 0.54

In [12]:
# Find the probabilities
transition_matrix = pd.crosstab(
    index=pitches_cleaned['previous_pitch_group'],
    columns=pitches_cleaned['next_pitch_group'],
    normalize='index'
)
transition_matrix = (transition_matrix * 100).round(1)
print("Pitch Sequencing Probabilities (Row = Previous Pitch, Column = Next Pitch):")
print(transition_matrix)

transition_matrix.to_csv('Pitch_Transition_Matrix.csv')

Pitch Sequencing Probabilities (Row = Previous Pitch, Column = Next Pitch):
next_pitch_group      Breaking  Fastball  Offspeed
previous_pitch_group                              
Breaking                  34.3      56.0       9.7
Fastball                  24.4      63.6      12.0
Offspeed                  19.1      55.9      25.0


Pitch Sequencing Probabilities

The probabilites displays a transition matrix showing the likelihood of the next pitch category based on the previous pitch category.

The Fastball First Rule: Regardless of what the previous pitch was a fastball is consistently the most probable next pitch. If the previous pitch was a Breaking ball there is a 56.0% chance the next pitch is a Fastball. If it was Offspeed there is a 55.9% chance of a Fastball to be next.

Pitch Repetition: Pitchers are highly likely to double down on fastballs. If a pitcher throws a fastball there is a 63.6% probability that the next pitch will also be a fastball. Breaking balls are repeated 34.3% of the time and Offspeed pitches are repeated 25.0% of the time.

Rarest Sequences: An Offspeed pitch following a Breaking pitch is the least likely sequence occurring only 9.7% of the time. Similarly an Offspeed pitch following a Fastball only happens 12% of the time.

Final interpretation: This is an excellent way to decide quickly what the probability of the next pitch will be. Although it doesn not predict the specific pitch it predicts a category but it tremendously helps batters to realize if they need to adjust to a quicker swing or a slower swing with better timing and swing accuracy.

In [13]:
importances = rf_grouped.feature_importances_
features = X_train.columns

feature_importance_df = pd.DataFrame({
    'Feature': features,
    'Importance_Score': importances
})

# Sort them from influential to least influential
feature_importance_df = feature_importance_df.sort_values(by='Importance_Score', ascending=False)

# Display the top 15 most critical game situations
print("Top 15 Pitch Prediction Drivers:")
print(feature_importance_df.head(15))



Top 15 Pitch Prediction Drivers:
                          Feature  Importance_Score
3                  pcount_pitcher          0.126144
15                     p_throws_R          0.104645
0                          inning          0.075379
8                        b_height          0.073872
14                        stand_R          0.067609
10                 home_team_runs          0.064435
9                  away_team_runs          0.063443
4                           balls          0.049618
2                   pcount_at_bat          0.039562
5                         strikes          0.038563
35  previous_pitch_group_Offspeed          0.035793
32         previous_pitch_type_SL          0.033880
7                            outs          0.032848
28         previous_pitch_type_KN          0.030487
6                           fouls          0.022887


Top Pitch Prediction Drivers

The top drivers reveals the features importances that are extracted from the model. This indicates which variables are carrying the most mathematical weight when the algorithm attempts to guess the next pitch.

The top 15 drivers can be grouped into the following conceptual categories:
Pitcher State & Physiology (Top Drivers):
pcount_pitcher (Score: 0.126): The pitchers total pitch count in the game is the number 1 most important variable. As a pitcher fatigues during the game their pitch selection changes drastically.

p_throws_R (Score: 0.105) & b_height (Score: 0.074): The pitchers handedness specifically if they are right-handed and the batter's height are critical as the ptichers can decide where in the xone they have to throw or if they have an advatange or disadvantage due to height.

In-Game Context:
The current inning (Score: 0.075) alongside the  game state—home_team_runs of (0.064) and away_team_runs (0.063)—highly dictate pitch strategy. Pitchers throw differently when leading versus trailing. This is shown to pitching aggressively, defensivley, or strategically.

At-Bat State (The Count):
The immediate micro-state of the duel is highly relevant. The number of balls (0.050)and the pitch count within that specific at-bat (pcount_at_bat, 0.040), strikes (0.039), outs (0.033), and fouls (0.023) all naturally steer the prediction. For example, a 3-0 count heavily skews the probability toward a fastball and can be used alongisde data from where in the strikezone it will most probably land.

Historical Sequencing Context:
Aligning with the transition matrix discussed previously the exact nature of the preceding pitch plays a role. Variables like previous_pitch_group_Offspeed (0.036), previous_pitch_type_SL (0.034) and previous_pitch_type_KN (0.030) show that what a pitcher just threw influences what they will throw next.

# XGBoost Model

In [ ]:
# Granular Setup for 17 Pitches
le_gran = LabelEncoder()
y_train_gran_num = le_gran.fit_transform(y_train_gran)
y_test_gran_num = le_gran.transform(y_test_gran)
weights_gran = compute_sample_weight(class_weight='balanced', y=y_train_gran_num)

# Grouped Setup 3 Categories
le_grp = LabelEncoder()
y_train_grp_num = le_grp.fit_transform(y_train_grp)
y_test_grp_num = le_grp.transform(y_test_grp)
weights_grp = compute_sample_weight(class_weight='balanced', y=y_train_grp_num)

# TRAIN MODEL for 17 PITCHES
xgb_gran = xgb.XGBClassifier(n_estimators=100, max_depth=6, learning_rate=0.1, random_state=42, n_jobs=-1)
xgb_gran.fit(X_train, y_train_gran_num, sample_weight=weights_gran)

# Predict and decode back to text
preds_gran_num = xgb_gran.predict(X_test)
preds_gran_text = le_gran.inverse_transform(preds_gran_num)

print("\nXGBOOST RESULTS 17 Pitches")
print("Accuracy:", accuracy_score(y_test_gran, preds_gran_text))
print(classification_report(y_test_gran, preds_gran_text))

# TRAIN MODEL for 3 CATEGORIES
xgb_grp = xgb.XGBClassifier(n_estimators=100, max_depth=6, learning_rate=0.1, random_state=42, n_jobs=-1)
xgb_grp.fit(X_train, y_train_grp_num, sample_weight=weights_grp)

# Predict and decode back to text
preds_grp_num = xgb_grp.predict(X_test)
preds_grp_text = le_grp.inverse_transform(preds_grp_num)

print("\nXGBOOST GROUPED RESULTS")
print("Accuracy:", accuracy_score(y_test_grp, preds_grp_text))
print(classification_report(y_test_grp, preds_grp_text))


# Features for grouped XGBoost model
xgb_importances = xgb_grp.feature_importances_

xgb_feature_df = pd.DataFrame({
    'Feature': X_train.columns,
    'Importance_Score': xgb_importances
}).sort_values(by='Importance_Score', ascending=False)

print("\nTop 15 XGBoost Pitch Prediction Drivers (Grouped Model):")
print(xgb_feature_df.head(15))

xgb_feature_df.to_csv('XGBoost_Pitch_Drivers.csv', index=False)


XGBOOST RESULTS 17 Pitches
Accuracy: 0.26215211589782694
              precision    recall  f1-score   support

          CH       0.21      0.27      0.24     11796
          CU       0.20      0.14      0.17      8112
          EP       0.00      0.11      0.00        27
          FC       0.34      0.23      0.27      6378
          FF       0.55      0.20      0.29     34218
          FS       0.06      0.33      0.10      1746
          FT       0.32      0.30      0.31     11437
          KC       0.06      0.22      0.10      1204
          KN       0.04      0.77      0.07       621
          SI       0.46      0.37      0.41     12248
          SL       0.31      0.33      0.32     17133

    accuracy                           0.26    104920
   macro avg       0.23      0.30      0.21    104920
weighted avg       0.38      0.26      0.29    104920


XGBOOST GROUPED RESULTS
Accuracy: 0.3851410598551277
              precision    recall  f1-score   support

    Breaking       0

17 Pitches

The dataset evaluates a broad spectrum of pitch types and they are categorized under the 17 Pitches model. Predicting exact pitch variations from such diverse options is incredibly difficult because pitchers actively employ deception, tunneling, and randomized sequencing to disrupt a hitter's timing. While traditional machine learning metrics might view the baseline accuracy as low, the model is successfully learning the underlying signals and pitcher tendencies rather than just randomly guessing across imbalanced data.

Grouped Results
When simplifying the classification into three broader categories (Fastball, Breaking, and Offspeed), the overall model accuracy increases to 38.5%. This grouped approach is highly practical because it mirrors how a hitter processes information in the batter's box. They often look for a general speed and movement of the ball or movement of the pitchers arm/hand rather than needing to identify the exact pitch type.
Fastball Precision: Fastballs comprise the majority of the dataset (64,281 pitches) and attain a highly actionable precision of 0.74. This is the most valuable metric because it means when the model predicts a fastball it is correct 74% of the time allowing a hitter to confidently expect a high speed ball.

Offspeed Recall: Offspeed pitches demonstrate the highest recall in this group at 0.63 though they have the lowest precision at 0.21.


Prediction Drivers
The mathematical features driving the XGBoost algorithm reveal that the model has effectively learned baseball psychology and game theory validating its situational awareness much like any player.

Sequencing Tactics: The most heavily weighted feature for the grouped model is "previous_pitch_group_Offspeed" at (0.228225). This perfectly illustrates the strategic pitching loop where an offspeed pitch is often followed by a fastball to exploit a batter's slowed down bat speed.

Physiological Matchups: The pitcher's handedness ("p_throws_R") is the second most important driver at (0.166765). This highlights the importance of advantages as the physical orientation of the pitcher dictates the sequence and arsenal used against specific batters.

Game State Context: The model intelligently factors in the macro and micro game states by utilizing immediate count variables like "strikes" and "balls," as well as the current "inning."

In [ ]:
# Encode targets into integers
label_encoder_gran = LabelEncoder()
y_train_num_gran = label_encoder_gran.fit_transform(y_train_gran)
y_test_num_gran = label_encoder_gran.transform(y_test_gran)
sample_weights_gran = compute_sample_weight(class_weight='balanced', y=y_train_num_gran)

# XGBoost Classifier
xgb_model_gran = xgb.XGBClassifier(
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1,
    random_state=42,
    n_jobs=-1
)

# Train model
xgb_model_gran.fit(X_train, y_train_num_gran, sample_weight=sample_weights_gran)

# Make predictions
preds_num_gran = xgb_model_gran.predict(X_test)
preds_text_gran = label_encoder_gran.inverse_transform(preds_num_gran)

# Evaluate the results
print("\nXGBOOST MODEL RESULTS")
print("Accuracy:", accuracy_score(y_test_gran, preds_text_gran))
print(classification_report(y_test_gran, preds_text_gran))


# Features importances from XGBoost
xgb_importances_gran = xgb_model_gran.feature_importances_

xgb_feature_df_gran = pd.DataFrame({
    'Feature': X_train.columns,
    'Importance_Score': xgb_importances_gran
}).sort_values(by='Importance_Score', ascending=False)

print("\nTop 15 XGBoost Granular Pitch Prediction Drivers:")
display(xgb_feature_df_gran.head(15))

xgb_feature_df_gran.to_csv('XGBoost_Granular_Pitch_Drivers.csv', index=False)


XGBOOST MODEL RESULTS
Accuracy: 0.26215211589782694
              precision    recall  f1-score   support

          CH       0.21      0.27      0.24     11796
          CU       0.20      0.14      0.17      8112
          EP       0.00      0.11      0.00        27
          FC       0.34      0.23      0.27      6378
          FF       0.55      0.20      0.29     34218
          FS       0.06      0.33      0.10      1746
          FT       0.32      0.30      0.31     11437
          KC       0.06      0.22      0.10      1204
          KN       0.04      0.77      0.07       621
          SI       0.46      0.37      0.41     12248
          SL       0.31      0.33      0.32     17133

    accuracy                           0.26    104920
   macro avg       0.23      0.30      0.21    104920
weighted avg       0.38      0.26      0.29    104920


Top 15 XGBoost Granular Pitch Prediction Drivers:


,Feature,Importance_Score
28,previous_pitch_type_KN,0.167438
27,previous_pitch_type_KC,0.091153
23,previous_pitch_type_FS,0.081880
31,previous_pitch_type_SI,0.081134
35,previous_pitch_group_Offspeed,0.070645
20,previous_pitch_type_FC,0.068279
24,previous_pitch_type_FT,0.043859
34,previous_pitch_group_Fastball,0.042581
32,previous_pitch_type_SL,0.035677
18,previous_pitch_type_EP,0.035527


Granular Results
The granular model attempts the difficult task of distinguishing between specific pitch types achieving an overall accuracy of 26.2%.

Four-Seam Dominance: Four-seam fastballs (FF) represent the largest portion of the data with 34,218 pitches showing a precision of 0.55 but a lower recall of 0.20.

Rare Pitches: The rarest pitch in the dataset is the Eephus (EP) which has only 27 occurrences. Due to its extreme rarity it results in a precision of 0.00 and a recall of 0.11.


Granular Predictions
When making highly specific predictions the granular model relies heavily on the exact type of the preceding pitch and successfully adapts to extreme anomalies in the dataset and this is good.

Preceding Pitch Reliance: The most dominant driver for granular predictions is "previous_pitch_type_KN" with an importance score of 0.167438, followed closely by other granular previous pitches like "KC" (0.091153), "FS" (0.081880) and "SI" (0.081134).

The Knuckleball Anomaly: The Knuckleball (KN) achieves the highest recall at a 0.77 despite a remarkably low precision of 0.04 across 621 instances. This reflects baseball reality knuckleballs are exceedingly rare but when they pitch they throw it almost exclusively. The model correctly learned to aggressively predict the knuckleball once it is introduced into a sequence.

**Final Decision**

The Random Forest model is the better choice for predicting the next pitch.

Superior Accuracy: A 41.6% accuracy in a 3-class baseball prediction model is notably strong given the randomness and many variables of the sport. It comfortably beats XGBoost's 38.5%.

Robust Feature Selection: Random Forest's reliance on a blend of fatigue (pcount), game context (inning), and matchups (stand_R, b_height) creates a more resilient model. XGBoost's heavy reliance on the immediate previous pitch makes it highly susceptible to being tricked or fooled if a pitcher breaks their typical sequencing pattern.

Actionable Balance: While XGBoost's 74% fastball precision is tempting, sacrificing that much recall limits its utility. Random Forest's 71% precision is still highly actionable for a hitter looking to expect on a something fast and it successfully identifies a much larger volume of those opportunities.